simulationdataschema/atoms_state.py

```diff
@@ -0,0 +1,571 @@
+        if self.umn is not None and self.orbitals_ref is not None:
+            if len(self.umn) != len(self.orbitals_ref):
+                logger.error(
+                    "The length of `HubbardInteractions.umn` does not coincide with length of `HubbardInteractions.orbitals_ref`."
+                )
+
+
+class AtomsState(ArchiveSection):
+    """
+    A base section to define each atom state information.
+    """
+
+    # ? constraint to the normal chemical elements (no 'X' as defined in ASE included)
+    chemical_symbol = Quantity(
+        type=MEnum(ase.data.chemical_symbols[1:]),
+        description="""
+        Symbol of the element, e.g. 'H', 'Pb'. This quantity is equivalent to `atomic_numbers`.
+        """,
+    )
+
+    atomic_number = Quantity(
+        type=np.int32,
+        description="""
+        Atomic number Z. This quantity is equivalent to `chemical_symbol`.
+        """,
+    )
+
+    orbitals_state = SubSection(sub_section=OrbitalsState.m_def, repeats=True)
+
+    charge = Quantity(
+        type=np.int32,
+        default=0,
+        description="""
+        Charge of the atom. It is defined as the number of extra electrons or holes in the
+        atom. If the atom is neutral, charge = 0 and the summation of all (if available) the`OrbitalsState.occupation`
+        coincides with the `atomic_number`. Otherwise, charge can be any positive integer (+1, +2...)
+        for cations or any negative integer (-1, -2...) for anions.
+
+        Note: for `CoreHole` systems we do not consider the charge of the atom even if
+        we do not store the final `OrbitalsState` where the electron was excited to.
+        """,
+        a_eln=ELNAnnotation(component="NumberEditQuantity"),
+    )
+
+    core_hole = SubSection(sub_section=CoreHole.m_def, repeats=False)
+
+    hubbard_interactions = SubSection(
+        sub_section=HubbardInteractions.m_def, repeats=False
+    )
+
+    def resolve_chemical_symbol_and_number(self, logger: BoundLogger) -> None:
+        """
+        Resolves the chemical symbol from the atomic number and viceversa.
+
+        Args:
+            logger (BoundLogger): The logger to log messages.
+        """
+        if self.chemical_symbol is None and self.atomic_number is not None:
+            try:
+                self.chemical_symbol = ase.data.chemical_symbols[self.atomic_number]
+            except IndexError:
+                logger.error(
+                    "The `AtomsState.atomic_number` is out of range of the periodic table."
+                )
+                return
+        elif self.chemical_symbol is not None and self.atomic_number is None:
+            try:
+                self.atomic_number = ase.data.atomic_numbers[self.chemical_symbol]
+            except IndexError:
+                logger.error(
+                    "The `AtomsState.chemical_symbol` is not recognized in the periodic table."
+                )
+                return
```

Comment:
Slight suggestion to improve legibility (which is anyhow already pretty good):

```
        f = lambda x: tuple(map(bool, x))
        combo = (self.chemical_symbol, self.atomic_number)

        if f(combo) == f((None, not None)):
            try:
                self.chemical_symbol = ase.data.chemical_symbols[self.atomic_number]
            except IndexError:
                logger.error(
                    "The `AtomsState.atomic_number` is out of range of the periodic table."
                )
        elif f(combo) == f((not None, None)):
            try:
                self.atomic_number = ase.data.atomic_numbers[self.chemical_symbol]
            except IndexError:
                logger.error(
                    "The `AtomsState.chemical_symbol` is not recognized in the periodic table."
                )
        return None
```

You could similarly use other tuples (or a dictionary for better semantics) to condense the mapping and error message too.

You might also prefer to return a boolean for error handling. Isn't necessary atm, but pays off down the road

Reply:
Done